File: 11.07/황지원/드래곤_커브.java
import java.io.*;
import java.util.*;

public class Main {

    static int N, X, Y, D, G;
    static ArrayList<ArrayList<Integer>> list = new ArrayList<>();
    static ArrayList<P> set = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            list = new ArrayList<>();
            st = new StringTokenizer(br.readLine());
            X = Integer.parseInt(st.nextToken());
            Y = Integer.parseInt(st.nextToken());
            D = Integer.parseInt(st.nextToken()) + 1;
            G = Integer.parseInt(st.nextToken());

            for (int k = 0; k <= G; k++) {
                list.add(new ArrayList<>());
            }

            draw();
        }

        int ans = 0;
        for (P p : set) {
            if ((rangeCheck(p.x, p.y) && rangeCheck(p.x, p.y + 1) && rangeCheck(p.x + 1, p.y) && rangeCheck(p.x + 1, p.y + 1))) {
                if (set.contains(new P(p.x, p.y + 1)) && 
                    set.contains(new P(p.x + 1, p.y)) && 
                    set.contains(new P(p.x + 1, p.y + 1))) {
                    ++ans;
                }
            }
        }

        System.out.println(ans);
    }

    static void draw() {
        // 1.list에 방향을 넣고
        // 2. 그 방향대로 이동한 후 끝점을 X, Y에 넣고
        // 3. set에 방문한 점을 넣어준다.
        if (!set.contains(new P(X, Y)))
            set.add(new P(X, Y)); 
        list.get(0).add(D); // 세대 방향
        moveXY(D); // XY이동
        if (!set.contains(new P(X, Y)))
            set.add(new P(X, Y)); // 방문 체크
        
        if (1 <= G) {
            list.get(1).add(D % 4 + 1); // 1세대 방향
            moveXY(D % 4 + 1); // XY이동
            if (!set.contains(new P(X, Y)))
                set.add(new P(X, Y)); // 방문 체크
        }

        for (int i = 2; i <= G; i++) {
            // 2세대 뒤에서부터
            for (int a = i - 1; a >= 0; a--) { // 1, 0
                for (int b = list.get(a).size() - 1; b >= 0; b--) {
                    int d = list.get(a).get(b) % 4 + 1;
                    list.get(i).add(d);
                    moveXY(d);
                    if (!set.contains(new P(X, Y)))
                        set.add(new P(X, Y)); // 방문 체크
                }
            }
        }
    }

    static void moveXY(int d) {
        if (d == 1) {
            X += 1;
        } else if (d == 2) {
            Y -= 1;
        } else if (d == 3) {
            X -= 1;
        } else if (d == 4) {
            Y += 1;
        }
    }

    static boolean rangeCheck(int x, int y) {
        return 0 <= x && x <= 100 && 0 <= y && y <= 100;
    }

    static class P {
        int x;
        int y;

        P(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        
        // 이거 없으면 객체 값 비교 안 됨 = contains 안됨
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            P p = (P) o;
            return x == p.x && y == p.y;
        }
    }
}

File: 2022/11.07/조수연/드래곤 커브.kt
package com.ssafy.algorithm

import java.io.*
import java.util.StringTokenizer

val direction = arrayOf(intArrayOf(1,0), intArrayOf(0,-1), intArrayOf(-1,0), intArrayOf(0,1)) //방향
var endX = 0
var endY = 0
lateinit var board : Array<IntArray>
var count = 0

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val N = br.readLine().toInt()
    board = Array(101){IntArray(101)}
    var dList = ArrayList<Int>()

    repeat(N){
        val token = StringTokenizer(br.readLine())

        val x = token.nextToken().toInt()
        val y = token.nextToken().toInt()
        val d = token.nextToken().toInt()
        val g = token.nextToken().toInt()

        dList.clear()
        board[y][x] = 1

        for (i in 0 .. g){
            when(i){
                0 ->  {
                    dList.add(d)
                    endX = x + direction[d][0]
                    endY = y + direction[d][1]
                    board[endY][endX] = 1

                }
                1 ->{
                    val curveDirection =  if (d - 1 < 0) 3 else d - 1
                    dList.add(0,curveDirection)
                    val nextX = endX - direction[curveDirection][0]
                    val nextY = endY - direction[curveDirection][1]
                    endX = nextX
                    endY = nextY
                    board[endY][endX] = 1
                }else -> {
                    curve(dList)
                }
            }
        }
    }

    //정사각형 찾기
    for (i in 0..100){
        for (j in 0..100){
            if (board[i][j] == 1){
                square(j,i)
            }
        }
    }

    println(count)

}

//드래곤 커브
fun curve(dList : ArrayList<Int>){

    val tempList = ArrayList<Int>()
    val startIndex = dList.size

    for (i in dList.size-1 downTo  0){ //방향 추가해주기
        val curveDirection =  if (dList[i] - 1 < 0) 3 else dList[i] - 1
        tempList.add(curveDirection) //다음 세대 드래곤 커브 방향 추가
    }

    dList.addAll(0,tempList)

    for (i in startIndex-1 downTo 0){ //방향이 끝 점 기준으로 처음에는 나가는 방향 그 다음은 들어오는 방향 규칙
        if(i % 2 != 0){ //끝 점을 기준으로 나가는 방향
            val nextX = endX + direction[dList[i]][0]
            val nextY = endY + direction[dList[i]][1]
            endX = nextX
            endY = nextY
        }else{ //끝 점을 기준으로 들어오는 방향
            val nextX = endX - direction[dList[i]][0]
            val nextY = endY - direction[dList[i]][1]
            endX = nextX
            endY = nextY
        }
        board[endY][endX] = 1
    }
}

//4점이 정사각형인지 체크
fun square(x : Int, y : Int){
    if (x + 1 < 101 && y + 1 < 101){ //범위체크
        if (board[y][x+1] == 1 && board[y+1][x] == 1 && board[y+1][x+1] == 1) count++ //세점이 1일때
    }
}

